src/main/groovy/liquibase/harness/compatibility/foundational/FoundationalCompatibilityTest.groovy

```diff
@@ -0,0 +1,97 @@
+package liquibase.harness.compatibility.foundational
+
+import liquibase.Scope
+import liquibase.database.jvm.JdbcConnection
+import liquibase.harness.config.DatabaseUnderTest
+import liquibase.harness.config.TestConfig
+import liquibase.harness.util.rollback.RollbackStrategy
+import liquibase.ui.UIService
+import spock.lang.Shared
+import spock.lang.Specification
+import spock.lang.Unroll
+
+import static liquibase.harness.compatibility.foundational.FoundationalCompatibilityTestHelper.buildTestInput
+import static liquibase.harness.util.TestUtils.chooseRollbackStrategy
+import static liquibase.harness.util.TestUtils.executeCommandScope
+
+@Unroll
+class FoundationalCompatibilityTest extends Specification {
+    @Shared
+    RollbackStrategy strategy;
+    @Shared
+    List<DatabaseUnderTest> databases
+    long timeMillisBeforeTest
+    long timeMillisAfterTest
+
+    def setupSpec() {
+        databases = TestConfig.instance.getFilteredDatabasesUnderTest()
+        strategy = chooseRollbackStrategy()
+        strategy.prepareForRollback(databases)
+    }
+
+    def "apply #testInput.change against #testInput.databaseName #testInput.version"() {
+        given: "read input data"
+        UIService uiService = Scope.getCurrentScope().getUI()
+
+        Map<String, Object> argsMap = new HashMap()
+        argsMap.put("url", testInput.url)
+        argsMap.put("username", testInput.username)
+        argsMap.put("password", testInput.password)
+        String baseChangelogPath = "liquibase/harness/compatibility/foundational/"
+        String insertChangelog = baseChangelogPath + "insert/" + testInput.change + ".xml"
+        String updateChangelog = testInput.baseChangelogPath.replace("setup", "update")
```

Comment:
this part is frustrating, why we use different approaches for different changetypes?

Reply:
++